App Submission: Fedimint

m1sterc001guy opened the pull request:
# App Submission

### Fedimint
...

### 256x256 SVG icon
![fedimint-icon-color-resized](https://github.com/user-attachments/assets/ac118a1e-8c01-4361-aaeb-18c6ee781b34)


### Gallery images
<img width="1400" height="900" alt="dashboard_formatted" src="https://github.com/user-attachments/assets/3af335f8-0c1b-426f-b65b-c334020673e6" />
<img width="1440" height="900" alt="dkg_formatted" src="https://github.com/user-attachments/assets/4120d2d5-628e-4050-b8e3-4026f24c53d2" />
<img width="1440" height="900" alt="login_formatted" src="https://github.com/user-attachments/assets/4b8d6979-b60f-4b74-9034-1fb85a758104" />
<img width="1440" height="900" alt="share_formatted" src="https://github.com/user-attachments/assets/568a3a58-9ec6-430e-ad9a-f9324c7b7485" />



### I have tested my app on:
- [X] umbrelOS on a Raspberry Pi
- [ ] umbrelOS on an Umbrel Home
- [X] umbrelOS on Linux VM

m1sterc001guy commented:
I think this is ready for review. We've been testing this on a Raspberry Pi 4 and promox Umbrel Linux VM and have setup a number of test federations. Let me know if any changes are necessary and if the images are sufficient!

bradleystachurski reviewed (approved):
Looks good!

In addition to the testing @m1sterc001guy did, I tested this using an Umbrel VM as part of the `v0.8.0` release (https://github.com/fedimint/fedimint/issues/7560).

al-lac commented on fedimintd/umbrel-app.yml:
```diff
@@ -0,0 +1,25 @@
+manifestVersion: 1
+id: fedimintd
```
It may make sense to just call it `fedimint`
```suggestion
id: fedimint
```

m1sterc001guy replied:
My original reasoning behind using `fedimintd` is that in the future we would also like to add additional Fedimint services to Umbrel, like the Lightning Gateway. I felt this was more clear and we did the same thing on the Start9 package too.

If you feel strongly though, I can change it.

m1sterc001guy replied:
Going to keep this as is unless there are strong feelings to change it.

al-lac reviewed (changes requested):
Hey @m1sterc001guy,

thanks for the submission!

I added some points that could or should be adapted.

Was not able to fully test it myself, tried to set it up, but at some point it failed. A good guide in the description would really help out here.

m1sterc001guy commented:
> can we test ourselves with a federation of 1 or do we need more federation members to test?

You can test by setting up a 1-1 federation but that won't test the network connections between peers. I would be happy to find time to setup a test federation with you guys if you want.

m1sterc001guy commented on fedimintd/docker-compose.yml:
```diff
@@ -0,0 +1,25 @@
+      - 8173:8173/udp # p2p iroh
+      - 8174:8174/udp # api iroh
+    volumes:
+      - ${APP_DATA_DIR}/data:/data
```
Added `data/.gitkeep` with it just being an empty file

m1sterc001guy commented on fedimintd/umbrel-app.yml:
```diff
@@ -0,0 +1,25 @@
+defaultUsername: ""
+defaultPassword: ""
+submitter: Fedimint Developers
+submission: https://github.com/getumbrel/umbrel-apps/pull/3371
```
Added newline

lukechilds commented:
@m1sterc001guy all good I did a test with a federation of 1 and connected to my iOS app and it worked flawlessly!

al-lac commented:
Yeah probably for the best to have it fixed already when we roll out the app. Any timeframe for `0.8.1`?

al-lac commented:
Hey again @m1sterc001guy @bradleystachurski, I now connected to my Fedimint that is running on my Pi 5 using the iOS app! 

When I am on the same network however, it sometimes connects and sometimes does not. I got this error when trying to add the federation:
<img width="1179" height="530" alt="image" src="https://github.com/user-attachments/assets/7cbaf070-16a0-41c4-a0e6-97111b84f72e" />

When I connected from outside the local network, it worked just fine. After reconnecting to my Wifi, I sometimes now seem to have a connection, sometimes not.

Probably just my local setup acting up, but just wanted to point it out here.

nmfretz commented on fedimintd/docker-compose.yml:
```diff
@@ -0,0 +1,25 @@
+    environment:
+      FM_ENABLE_IROH: "true"
+      FM_DATA_DIR: "/data"
+      FM_BITCOIN_NETWORK: bitcoin
```
**TLDR:** change this to:
```suggestion
      FM_BITCOIN_NETWORK: "${APP_BITCOIN_NETWORK_ELECTRS}"
```

and we get fedimintd able to work on all Bitcoin networks for free, not just mainnet!

----------------------

All Bitcoin node apps on umbrelOS (Bitcion Node (Core), Knots, Libre Relay) can be easily switched to any network by a user within the app's settings, and it then exports an environment variable for connected apps to use to know what network to run on. So a cool thing we can do here is set `FM_BITCOIN_NETWORK` dynamically so that whatever network a user has their connected Bitcoin node running on will be passed through here. So say you're running the Bitcoin Node app on signet, you could install fedimintd and it would work out of the box on signet. 

I see from https://github.com/fedimint/fedimint/blob/master/fedimintd/src/lib.rs that fedimintd is using the `Network` type from the rust bitcoin library (https://docs.rs/bitcoin/latest/bitcoin/network/enum.Network.html). The bitcoin apps actually already export a compatible variable called `APP_BITCOIN_NETWORK_ELECTRS` which was originally for the Electrs app.

I'll test to make sure this is working well.

m1sterc001guy replied:
Fixed!

nmfretz commented on fedimintd/umbrel-app.yml:
```diff
@@ -0,0 +1,66 @@
+
+  # Enabling Lightning Payments
+
+  Fedimint supports Lightning payments via a Lightning Gateway. To use Lightning in your federation, there are two options:
+
+  1. Use an existing Lightning Gateway. [Reach out on Discord to get setup](https://chat.fedimint.org).
+
+  2. Run your own - [see Lightning Gateway docs for more information](https://github.com/fedimint/fedimint/blob/master/docs/gateway.md).
```
The plan on umbrelOS eventually is to have "optional dependencies".  e.g if this app exists, I want to integrate with it, but if this app doesn't exist I can still function without it.

It would be cool in the future to have it so that a user can choose to use their own LND instance on their Umbrel as a Gateway and we take care of the initial config (like we do here already for bitcoind).

nmfretz replied:
Nice! Excited to tinker and help set that up whenever it's ready.

nmfretz commented:
Tested on x86 (awaiting arm fix to test on the Pi) and working great guys! Super easy set-up, really great work.
- Tested x86
- Tested connection via iOS app on both local and remote network.
- Tested compatibility with the three Bitcoin node implementations currently on the app store
- Tested that dynamic network env var works (`FM_BITCOIN_NETWORK: ${APP_BITCOIN_NETWORK_ELECTRS}`). Specifically tested running the Bitcoin Node app on `signet`.

Here's a couple small hiccups I noticed. These are totally non-blocking for getting this to the app store, but I'm just noting them here in case you haven't run across them:

1) copy to clipboard functionality won't work on http
This is for the "Invite Code" copy to clipboard buttons. Not critical since users can just select and copy manually.

2) After I clicked to move to the DKG step, the automatic redirect gave me a connection error for the UI. After hard refreshing everything was fine. I didn't grab container, browser console, or network logs so I'll try to recreate it and grab them next time.

nmfretz commented on fedimintd/umbrel-app.yml:
```diff
@@ -0,0 +1,66 @@
+
+  # Step By Step Ceremony
+
+  1. Install & Launch fedimintd
+     Each participating guardian runs their own instance of fedimintd.
```
Sweet, thanks

m1sterc001guy commented:
Hey guys! [Fedimint v0.8.1 has been released](https://github.com/fedimint/fedimint/releases/tag/v0.8.1), so I have updated this PR with the new image and incorporated the remaining feedback. 

I have tested on my own Raspberry Pi 4 and it is working. To re-create the 64KB page issue on arm, I used a VM in the cloud that I verified reproduced the issue on `v0.8.0`. `v0.8.1` is able to start successfully on the same VM.

 Let me know if there's additional testing that should be done.

al-lac commented:
Hey @m1sterc001guy! Nice! Great job on fixing this so quickly.

Also tested it on my Raspberry Pi 5 and it works flawlessly now 🚀 

Waiting on the final review of @nmfretz and the app store assets, then we should be ready to go live.

nmfretz commented:
Excellent work @m1sterc001guy! Reviewed again and tested. Once the app store assets are done we'll go live!

m1sterc001guy commented:
@nmfretz They look good to me! Thanks for making those. Some small suggestions:

 There's a typo in the last photo: `Fediment -> Fedimint`.

I like the photo that showcases the Fedi app, it does give users an idea of how to interact with it. They are a separate company from the Fedimint open source project though, and they are other wallets like [Harbor](https://github.com/HarborWallet/harbor) and [Vipr](https://github.com/ngutech21/vipr-wallet) that can also interact with Fedimint (with another [wallet soon to be released](https://github.com/fedimint/e-cash-app)). Could we maybe add screenshots of these apps too?

nmfretz commented:
@m1sterc001guy whoops, great catches on the last image! Ya, we'll fix that to showcase several apps and change the text to be more generic... something like "Allow other to connect from wallets that support Fedimint".

nmfretz commented:
Alrighty @m1sterc001guy we're going live 🎉 Thanks a ton for bringing Fedimint to Umbrel! This is such a great use-case.

We've updated the 3rd gallery shot (see updated comment above).

Here's what it will look like in the app store:
<img width="1783" height="1104" alt="image" src="https://github.com/user-attachments/assets/e310dbed-7fed-4cdc-8cf2-cd3bd0debc0e" />

### App Updates

Whenever you have an update you can open a PR to this repo again and we will help review and test. The main changes needed for all updates are:
- Updated image in the docker-compose.yml.
- Updated `version` and `releaseNotes` values in the umbrel-app.yml

The updated `version` field is what triggers umbrelOS to show users that there is an available app update for them.

Here's a diff from a Bitcoin Node app update as an example: https://github.com/getumbrel/umbrel-apps/pull/3123/files

### Badges

If at some point you're looking for material to show you're on the Umbrel app store, feel free to use these badges that are dynamically generated by our app store preview site:

The general url for an app-specific badge is:
```
https://apps.umbrel.com/api/app/<app-id>/badge-<light-or-dark>.svg
```

So, for Fedimint:

Dark badge:
![app-dark-badge](https://apps.umbrel.com/api/app/fedimintd/badge-dark.svg)

Light badge:
![app-light-badge](https://apps.umbrel.com/api/app/fedimintd/badge-light.svg)

Or you can also just use a generic, non-app-specific badge:
```
https://apps.umbrel.com/badge-<light-or-dark>.svg
```

Dark:
![dark-badge](https://apps.umbrel.com/badge-dark.svg)

Light:
![light-badge](https://apps.umbrel.com/badge-light.svg)

m1sterc001guy commented:
Thank you guys! Looks amazing. Appreciate all the help getting over the finish line.
